src/jit/compiler.cpp

```diff
@@ -2321,58 +2321,42 @@ const char* Compiler::compLocalVarName(unsigned varNum, unsigned offs)
 #ifdef _TARGET_XARCH_
 static bool configEnableISA(InstructionSet isa)
 {
-#ifdef DEBUG
     switch (isa)
     {
         case InstructionSet_SSE:
             return JitConfig.EnableSSE() != 0;
         case InstructionSet_SSE2:
-            return JitConfig.EnableSSE2() != 0;
+            return JitConfig.EnableSSE() != 0 && JitConfig.EnableSSE2() != 0;
         case InstructionSet_SSE3:
-            return JitConfig.EnableSSE3() != 0;
+            return configEnableISA(InstructionSet_SSE2) && JitConfig.EnableSSE3() != 0;
         case InstructionSet_SSSE3:
-            return JitConfig.EnableSSSE3() != 0;
+            return configEnableISA(InstructionSet_SSE3) && JitConfig.EnableSSSE3() != 0;
         case InstructionSet_SSE41:
-            return JitConfig.EnableSSE41() != 0;
+            return configEnableISA(InstructionSet_SSSE3) && JitConfig.EnableSSE41() != 0;
         case InstructionSet_SSE42:
-            return JitConfig.EnableSSE42() != 0;
+            return configEnableISA(InstructionSet_SSE41) && JitConfig.EnableSSE42() != 0;
         case InstructionSet_AVX:
-            return JitConfig.EnableAVX() != 0;
+            return configEnableISA(InstructionSet_SSE42) && JitConfig.EnableAVX() != 0;
         case InstructionSet_FMA:
-            return JitConfig.EnableFMA() != 0;
+            return configEnableISA(InstructionSet_AVX) && JitConfig.EnableFMA() != 0;
         case InstructionSet_AVX2:
-            return JitConfig.EnableAVX2() != 0;
+            return configEnableISA(InstructionSet_AVX) && JitConfig.EnableAVX2() != 0;
 
         case InstructionSet_AES:
-            return JitConfig.EnableAES() != 0;
+            return configEnableISA(InstructionSet_SSE2) && JitConfig.EnableAES() != 0;
         case InstructionSet_BMI1:
-            return JitConfig.EnableBMI1() != 0;
+            return configEnableISA(InstructionSet_AVX) && JitConfig.EnableBMI1() != 0;
```

Comment:
We can't have the AVX check here because it will block BMI1/BMI2 in the scenario where the hardware supports it but the OS has disabled the YMM state support (as can be done using `bcdedit`, for example).

I think we determined that we need to have a new flag, `COMPlus_EnableVEX`, which controls whether or not the VEX encoding was supported. `COMPlus_EnableAVX` would continue being retail and would continue exclusively controlling AVX+ support (which maintains the externally observable functional behavior for the `System.Numerics.Vector` types).
* It (EnableVEX) likely does not need to be retail, as it only impacts the encoding of the instructions and shouldn't have any functional impact outside CoreCLR. The one caveat being that it does impact whether `LoadAligned` (non-VEX) or `LoadUnaligned` (VEX) is folded into the encoding of the next instruction (where-applicable), but I don't think that should be a major concern for the majority of applications.

Reply:
>  it will block BMI1/BMI2 in the scenario where the hardware supports it but the OS has disabled the YMM state support (as can be done using bcdedit, for example).

Right, but this statement should read "it will block **_testing_** BMI1/BMI2 in the scenario where the hardware supports it but the OS has disabled the YMM state support". 

This is a very very rare situation that .NET Core runs on a new hardware with old OS. I am not sure if it worthwhile to add a new flag (EnableVEX), which may depend on the test matrix and target environment concerns.

Reply:
> Right, but this statement should read "it will block testing...

Once these are retail flags, it will block the use of BMI1/BMI2 in production scenarios when the user disables AVX support. Wanting to test 128-bit SSE-SSE4.2 code that coordinates with BMI1/2 code is a completely valid scenario and you should be able to disable the AVX intrinsics separately from the VEX encoding support. This is especially valid when you consider that just using 256-bit registers is not always a clear cut improvement over using 128-bit registers, depending on the size of the payload, the exact instructions being executed, etc.

> This is a very very rare situation that .NET Core runs on a new hardware with old OS.

This does not just impact new hardware on an old OS. Windows has the built in capability for the user to disable the YMM state support using `bcdedit /set xsavedisable 1` (this includes Windows 10).